src/psyclone/psyir/nodes/omp_task_directive.py

```diff
@@ -0,0 +1,2136 @@
+            else:
+                private_clause.addchild(ref)
+        for ref in firstprivate_list:
+            firstprivate_clause.addchild(ref)
+        shared_clause = OMPSharedClause()
+        for ref in shared_list:
+            shared_clause.addchild(ref)
+
+        in_clause = OMPDependClause(
+            depend_type=OMPDependClause.DependClauseTypes.IN
+        )
+        # For input references we need to ignore references to constant
+        # symbols. This means we need to try to get external symbols as well
+        for ref in in_list:
+            if isinstance(ref.symbol, DataSymbol) and ref.symbol.is_constant:
+                continue
+            if (ref.symbol.is_import and
+                    ref.symbol.get_external_symbol().is_constant):
+                continue
+            in_clause.addchild(ref)
+        out_clause = OMPDependClause(
+            depend_type=OMPDependClause.DependClauseTypes.OUT
+        )
+        for ref in out_list:
+            out_clause.addchild(ref)
+
+        return (
+            private_clause,
+            firstprivate_clause,
+            shared_clause,
+            in_clause,
+            out_clause,
+        )
+
+    def lower_to_language_level(self):
+        """
+        Lowers the structure of the PSyIR tree inside the Directive
+        to generate the Clauses that are required for this Directive.
+        """
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyGen import Kern
+
+        # If we find a Kern or Call child then we abort.
+        # Note that if the transformation is used it will have already
+        # attempted to do this inlining.
+        if len(self.walk((Kern, Call))) > 0:
+            raise GenerationError(
+                "Attempted to lower to OMPTaskDirective "
+                "node, but the node contains a Call or Kern "
+                "which must be inlined first."
+            )
+
+        # Create the clauses
+        (
+            private_clause,
+            firstprivate_clause,
+            shared_clause,
+            in_clause,
+            out_clause,
+        ) = self._compute_clauses()
+
+        if len(self.children) < 2 or private_clause != self.children[1]:
+            self.children[1] = private_clause
+        if len(self.children) < 3 or firstprivate_clause != self.children[2]:
+            self.children[2] = firstprivate_clause
+        if len(self.children) < 4 or shared_clause != self.children[3]:
+            self.children[3] = shared_clause
+        if len(self.children) < 5 or in_clause != self.children[4]:
+            self.children[4] = in_clause
+        if len(self.children) < 6 or out_clause != self.children[5]:
+            self.children[5] = out_clause
+        super().lower_to_language_level()
+
+        # Replace this node with an OMPTaskDirective
+        childs = self.pop_all_children()
+        replacement = OMPTaskDirective(children=childs, lowering=True)
```

Comment:
I have to admit I'm not keen on the 'lowering=True' argument to the constructor. In fact, I'm not sure it's necessary since you do `child=self.pop_all_children()` immediately before it anyway?

Reply:
I think its still different inside `OMPTaskDirective`. The call at 2134 results in:
`childs = [Schedule, OMPClause1, OMPClause2, ...]`.
Inside the lowered version, we do `pop_all_children` on the first child, which gives us all the children of the `schedule`, which is what we need to call the `Node` `__init__` call. Does that make sense? I could instead try and do something like
`children=None, parent=None, clauses=None` and pass the children of the schedule as `children` and the clauses as a separate list if you think that would be better?

Reply:
Much nicer now, thanks.